File: packages/flags-flutter/test/exposure_test.dart
/// Verifies exposure-event behavior, especially that `'targeted'` rules
/// SKIP `$exposure` events. This is a hard cross-SDK requirement.
library;

import 'dart:async';
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:signakit_flags/signakit_flags.dart';

const String _testSdkKey = 'sk_dev_org123_proj456_random';

Map<String, Object?> _flagJson({
  required String key,
  required List<Map<String, Object?>> rules,
  required List<Map<String, Object?>> defaultRanges,
}) {
  return <String, Object?>{
    'id': '1',
    'key': key,
    'salt': 'salt-$key',
    'status': 'active',
    'running': true,
    'variations': <Map<String, Object?>>[
      <String, Object?>{'key': 'control'},
      <String, Object?>{'key': 'treatment'},
      <String, Object?>{'key': 'off'},
    ],
    'allocation': <String, Object?>{'ranges': defaultRanges},
    'rules': rules,
  };
}

Map<String, Object?> _ruleJson({
  required String key,
  required String type,
  required int trafficPercentage,
  required String variation,
}) {
  return <String, Object?>{
    'ruleKey': key,
    'ruleType': type,
    'trafficPercentage': trafficPercentage,
    'variationAllocation': <String, Object?>{
      'ranges': <Map<String, Object?>>[
        <String, Object?>{'variation': variation, 'start': 0, 'end': 9999},
      ],
    },
  };
}

Map<String, Object?> _projectConfig(List<Map<String, Object?>> flags) {
  return <String, Object?>{
    'projectId': 'proj456',
    'environmentKey': 'development',
    'sdkKey': _testSdkKey,
    'version': 1,
    'flags': flags,
    'generatedAt': '2025-01-01T00:00:00Z',
  };
}

/// Tracks all event POSTs.
class _RecordingMock {
  final List<Map<String, Object?>> events = <Map<String, Object?>>[];
  final List<http.Request> requests = <http.Request>[];

  http.Client client(Map<String, Object?> config) {
    return MockClient((http.Request request) async {
      requests.add(request);

      if (request.method == 'GET' &&
          request.url.path.endsWith('/latest.json')) {
        return http.Response(
          jsonEncode(config),
          200,
          headers: <String, String>{
            'content-type': 'application/json',
            'etag': '"v1"',
          },
        );
      }

      if (request.method == 'POST' &&
          request.url.toString().contains('flag-events')) {
        final body = jsonDecode(request.body) as Map<String, Object?>;
        final list = (body['events']! as List<Object?>);
        for (final e in list) {
          events.add(e! as Map<String, Object?>);
        }
        return http.Response('{}', 200);
      }

      return http.Response('not found', 404);
    });
  }
}

void main() {
  group('exposure events', () {
    test('targeted rule does NOT fire \$exposure', () async {
      final mock = _RecordingMock();
      final config = _projectConfig(<Map<String, Object?>>[
        _flagJson(
          key: 'targeted-flag',
          rules: <Map<String, Object?>>[
            _ruleJson(
              key: 'rollout',
              type: 'targeted',
              trafficPercentage: 100,
              variation: 'treatment',
            ),
          ],
          defaultRanges: <Map<String, Object?>>[
            <String, Object?>{
              'variation': 'control',
              'start': 0,
              'end': 9999,
            },
          ],
        ),
      ]);

      final client = SignaKitClient(SignaKitClientConfig(
        sdkKey: _testSdkKey,
        httpClient: mock.client(config),
      ));
      final ready = await client.onReady();
      expect(ready.success, isTrue);

      final ctx = client.createUserContext('user-123');
      expect(ctx, isNotNull);

      final decision = ctx!.decide('targeted-flag');
      expect(decision, isNotNull);
      expect(decision!.ruleType, equals(RuleType.targeted));
      expect(decision.variationKey, equals('treatment'));

      // Yield to let any scheduled futures run.
      await Future<void>.delayed(const Duration(milliseconds: 50));

      // No exposure event should have been posted.
      final exposureEvents = mock.events
          .where((e) => e['eventKey'] == r'$exposure')
          .toList(growable: false);
      expect(
        exposureEvents,
        isEmpty,
        reason: 'targeted rules must skip \$exposure tracking',
      );

      client.close();
    });

    test('ab-test rule DOES fire \$exposure', () async {
      final mock = _RecordingMock();
      final config = _projectConfig(<Map<String, Object?>>[
        _flagJson(
          key: 'ab-flag',
          rules: <Map<String, Object?>>[
            _ruleJson(
              key: 'experiment',
              type: 'ab-test',
              trafficPercentage: 100,
              variation: 'treatment',
            ),
          ],
          defaultRanges: <Map<String, Object?>>[
            <String, Object?>{
              'variation': 'control',
              'start': 0,
              'end': 9999,
            },
          ],
        ),
      ]);

      final client = SignaKitClient(SignaKitClientConfig(
        sdkKey: _testSdkKey,
        httpClient: mock.client(config),
      ));
      await client.onReady();

      final ctx = client.createUserContext('user-abc');
      final d = ctx!.decide('ab-flag');
      expect(d!.ruleType, equals(RuleType.abTest));

      await Future<void>.delayed(const Duration(milliseconds: 50));

      final exposures =
          mock.events.where((e) => e['eventKey'] == r'$exposure').toList();
      expect(exposures, hasLength(1));
      final exp = exposures.first;
      expect(exp['userId'], equals('user-abc'));
      final decisions = exp['decisions']! as Map<String, Object?>;
      expect(decisions['ab-flag'], equals('treatment'));
      final metadata = exp['metadata']! as Map<String, Object?>;
      expect(metadata['flagKey'], equals('ab-flag'));
      expect(metadata['ruleKey'], equals('experiment'));

      client.close();
    });

    test('bot user agent → off, no events', () async {
      final mock = _RecordingMock();
      final config = _projectConfig(<Map<String, Object?>>[
        _flagJson(
          key: 'ab-flag',
          rules: <Map<String, Object?>>[
            _ruleJson(
              key: 'experiment',
              type: 'ab-test',
              trafficPercentage: 100,
              variation: 'treatment',
            ),
          ],
          defaultRanges: <Map<String, Object?>>[
            <String, Object?>{'variation': 'control', 'start': 0, 'end': 9999},
          ],
        ),
      ]);

      final client = SignaKitClient(SignaKitClientConfig(
        sdkKey: _testSdkKey,
        httpClient: mock.client(config),
      ));
      await client.onReady();

      final ctx = client.createUserContext(
        'bot-user',
        attributes: <String, Object?>{
          r'$userAgent':
              'Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)',
        },
      );

      final d = ctx!.decide('ab-flag');
      expect(d!.variationKey, equals('off'));
      expect(d.enabled, isFalse);

      await ctx.trackEvent('purchase', value: 9.99);
      await Future<void>.delayed(const Duration(milliseconds: 50));

      expect(mock.events, isEmpty);
      client.close();
    });

    test('parseSdkKey handles dev/prod', () {
      // Smoke test that the SDK key parses through the public client.
      final mock = _RecordingMock();
      final config = _projectConfig(const <Map<String, Object?>>[]);

      final c1 = SignaKitClient(SignaKitClientConfig(
        sdkKey: 'sk_prod_org_proj_random',
        httpClient: mock.client(config),
      ));
      expect(c1, isNotNull);

      expect(
        () => SignaKitClient(SignaKitClientConfig(
          sdkKey: 'invalid',
          httpClient: mock.client(config),
        )),
        throwsArgumentError,
      );

      c1.close();
    });
  });
}

File: packages/flags-react-native/__tests__/exposure.test.ts
/**
 * Exposure POST behavior tests.
 *
 * Verifies the hard requirement: when `decision.ruleType === 'targeted'`,
 * the SDK MUST NOT fire a `$exposure` event.
 */

import { SignaKitClient } from '../src/client'
import { SIGNAKIT_EVENTS_URL } from '../src/constants'
import type { ProjectConfig } from '../src/types'

function makeConfig(): ProjectConfig {
  return {
    projectId: '1',
    environmentKey: 'production',
    sdkKey: 'sk_prod_org_1_xyz',
    version: 1,
    generatedAt: new Date().toISOString(),
    flags: [
      // Targeted rule: should NOT fire exposure.
      {
        id: 'tf',
        key: 'targeted-flag',
        salt: 's-targeted',
        status: 'active',
        running: true,
        variations: [{ key: 'on' }, { key: 'off' }],
        allocation: { ranges: [{ variation: 'off', start: 0, end: 9999 }] },
        rules: [
          {
            ruleKey: 'r-tgt',
            ruleType: 'targeted',
            trafficPercentage: 100,
            variationAllocation: { ranges: [{ variation: 'on', start: 0, end: 9999 }] },
          },
        ],
      },
      // A/B test rule: SHOULD fire exposure.
      {
        id: 'af',
        key: 'ab-flag',
        salt: 's-ab',
        status: 'active',
        running: true,
        variations: [{ key: 'on' }, { key: 'off' }],
        allocation: { ranges: [{ variation: 'off', start: 0, end: 9999 }] },
        rules: [
          {
            ruleKey: 'r-ab',
            ruleType: 'ab-test',
            trafficPercentage: 100,
            variationAllocation: { ranges: [{ variation: 'on', start: 0, end: 9999 }] },
          },
        ],
      },
    ],
  }
}

type FetchCall = { url: string; method: string; body: unknown }

function installFetchMock(configFetchUrlMatcher: (url: string) => boolean): {
  calls: FetchCall[]
  restore: () => void
} {
  const calls: FetchCall[] = []
  const original = (global as unknown as { fetch?: typeof fetch }).fetch
  const config = makeConfig()

  const mock = jest.fn(async (url: string, init?: RequestInit) => {
    const method = (init?.method ?? 'GET').toUpperCase()
    let parsedBody: unknown = undefined
    if (init?.body && typeof init.body === 'string') {
      try {
        parsedBody = JSON.parse(init.body)
      } catch {
        parsedBody = init.body
      }
    }
    calls.push({ url, method, body: parsedBody })

    // Config CDN GET
    if (configFetchUrlMatcher(url)) {
      return {
        ok: true,
        status: 200,
        statusText: 'OK',
        headers: { get: (_h: string) => null },
        json: async () => config,
      } as unknown as Response
    }

    // Events POST
    if (url === SIGNAKIT_EVENTS_URL) {
      return {
        ok: true,
        status: 200,
        statusText: 'OK',
        headers: { get: (_h: string) => null },
        json: async () => ({}),
      } as unknown as Response
    }

    return {
      ok: false,
      status: 404,
      statusText: 'Not Found',
      headers: { get: (_h: string) => null },
      json: async () => ({}),
    } as unknown as Response
  })
  ;(global as unknown as { fetch: unknown }).fetch = mock as unknown as typeof fetch

  return {
    calls,
    restore: () => {
      ;(global as unknown as { fetch?: typeof fetch }).fetch = original
    },
  }
}

describe('exposure tracking', () => {
  let calls: FetchCall[]
  let restore: () => void

  beforeEach(() => {
    const m = installFetchMock((u) => u.includes('/configs/'))
    calls = m.calls
    restore = m.restore
  })

  afterEach(() => {
    restore()
  })

  test('targeted-rule decisions do NOT fire $exposure', async () => {
    const client = new SignaKitClient({ sdkKey: 'sk_prod_org_1_xyz' })
    const ready = await client.onReady()
    expect(ready.success).toBe(true)

    const ctx = client.createUserContext('user-1')
    const decision = ctx?.decide('targeted-flag')
    expect(decision?.ruleType).toBe('targeted')
    expect(decision?.variationKey).toBe('on')

    // Allow microtasks to drain so any (incorrect) fire-and-forget fetch
    // would have been recorded.
    await Promise.resolve()
    await Promise.resolve()

    const exposurePosts = calls.filter(
      (c) => c.url === SIGNAKIT_EVENTS_URL && c.method === 'POST'
    )
    expect(exposurePosts).toHaveLength(0)
  })

  test('ab-test decisions DO fire a $exposure POST', async () => {
    const client = new SignaKitClient({ sdkKey: 'sk_prod_org_1_xyz' })
    await client.onReady()

    const ctx = client.createUserContext('user-1')
    const decision = ctx?.decide('ab-flag')
    expect(decision?.ruleType).toBe('ab-test')

    // Wait for the fire-and-forget fetch to actually be invoked.
    await new Promise((r) => setTimeout(r, 0))

    const exposurePosts = calls.filter(
      (c) => c.url === SIGNAKIT_EVENTS_URL && c.method === 'POST'
    )
    expect(exposurePosts).toHaveLength(1)
    const body = exposurePosts[0]!.body as { events: Array<{ eventKey: string; metadata: { flagKey: string; ruleKey: string } }> }
    expect(body.events[0]!.eventKey).toBe('$exposure')
    expect(body.events[0]!.metadata.flagKey).toBe('ab-flag')
    expect(body.events[0]!.metadata.ruleKey).toBe('r-ab')
  })

  test('exposure dedupes per session (same flag + same user)', async () => {
    const client = new SignaKitClient({ sdkKey: 'sk_prod_org_1_xyz' })
    await client.onReady()

    const ctx = client.createUserContext('user-1')
    ctx?.decide('ab-flag')
    ctx?.decide('ab-flag')
    ctx?.decide('ab-flag')

    await new Promise((r) => setTimeout(r, 0))

    const exposurePosts = calls.filter(
      (c) => c.url === SIGNAKIT_EVENTS_URL && c.method === 'POST'
    )
    expect(exposurePosts).toHaveLength(1)
  })
})
